src/runtime/HalideBuffer.h

```diff
@@ -1593,7 +1610,221 @@ class Buffer {
+                }
+                // Note we don't delete ptr->ptr. If it refers to
+                // a user Buffer then we don't own the memory. If
+                // it refers to ptr->storage then we're already
+                // deleting that.
+                delete ptr;
+                ptr = nullptr;
+            }
+        }
+    }
+
+    template<typename T2, int D2>
+    void assert_can_convert_from() {
+        static_assert(D <= D2,
+                      "A Buffer::Ref may only be constructed from Buffer::Ref of greater dimensionality");
+        static_assert(!std::is_const<T2>::value || std::is_const<T>::value,
+                      "A Buffer<T>::Ref may not be constructed from a Buffer<const T>::Ref");
+        static_assert(std::is_same<typename std::remove_const<T>::type,
+                      typename std::remove_const<T2>::type>::value ||
+                      std::is_same<typename std::remove_const<T>::type, void>::value,
+                      "A Buffer<T1>::Ref may only be constructed from a Buffer<T2>::Ref if T1 = T2 or T1 = void");
+    }
+
+    template<typename T2, int D2> friend class BufferRef;
+
+public:
+
+    /** Make a null BufferRef, which points to no Buffer */
+    BufferRef() {}
+
+    /** This constructor is only used by Buffer::make_shared_ref() */
+    BufferRef(typename Buffer<T, D>::RefHolder *r) : ptr(r) {
+        incref();
+    }
+
+    /** The conventional constructors, destructors, and assignment
+     * operators for a shared-pointer class. */
+    // @{
+    BufferRef(const BufferRef<T, D> &other) {
+        ptr = other.ptr;
+        incref();
+    }
+
+    template<typename T2, int D2>
+    BufferRef(const BufferRef<T2, D2> &other) {
+        assert_can_convert_from<T2, D2>();
+        ptr = other.ptr;
+        incref();
+    }
+
+    BufferRef(BufferRef<T, D> &&other) {
+        ptr = other.ptr;
+        other.ptr = nullptr;
+    }
+
+    template<typename T2, int D2>
+    BufferRef(BufferRef<T2, D2> &&other) {
+        assert_can_convert_from<T2, D2>();
+        ptr = (typename Buffer<T, D>::RefHolder *)other.ptr;
+        other.ptr = nullptr;
+    }
+
+    ~BufferRef() {
+        decref();
+    }
+
+    BufferRef &operator=(const BufferRef<T, D> &other) {
+        if (other.ptr != ptr) {
+            decref();
+            ptr = other.ptr;
+            incref();
+        }
+        return *this;
+    }
+
+    template<typename T2, int D2>
+    BufferRef &operator=(const BufferRef<T2, D2> &other) {
+        assert_can_convert_from<T2, D2>();
+        auto *other_ptr = (typename Buffer<T, D>::RefHolder *)other.ptr;
+        if (other_ptr != ptr) {
+            decref();
+            ptr = other_ptr;
+            incref();
+        }
+        return *this;
+    }
+
+    BufferRef &operator=(const BufferRef<T, D> &&other) {
+        if (other.ptr != ptr) {
+            decref();
+            ptr = other.ptr;
+            other.ptr = nullptr;
```

Comment:
I think maybe these 3 things should happen unconditionally?

I think the current implementation may technically be correct, but might increase memory usage in a corner case usage when the ref being moved is the same.